### PR TITLE
unlock add

### PR DESCRIPTION
**Pull request by marian-craciunescu**

### CODE REVIEW

**Comment by zeisss** [comment]
LGTM

File: server/connector/manager.go [hunk 1]
```diff
@@ -168,5 +200,7 @@ func (m *manager) updateStore(s Subscriber) error {
 }
 
 func (m *manager) removeStore(s Subscriber) error {
+	//TODO MARIAN also remove this logs.
+	logger.WithField("subscriber", s).Debug("RemoveStore")
```
Comment (cosminrentea) [inline server/connector/manager.go]: maybe remove

File: server/connector/manager.go [hunk 2]
```diff
@@ -144,19 +171,24 @@ func (m *manager) Exists(key string) bool {
...
+	if err != nil {
+		return err
 	}
+	m.deleteSubscriberFromMap(s)
 
-	delete(m.subscribers, s.Key())
 	return m.removeStore(s)
 }
 
+func (m *manager) cancelSubscribers(s Subscriber) {
```
Comment (cosminrentea) [inline server/connector/manager.go]: cancelSubscriber

File: server/connector/manager.go [hunk 3]
```diff
@@ -106,33 +104,62 @@ func (m *manager) Filter(filters map[string]string) (subscribers []Subscriber) {
...
+	defer m.RUnlock()
+
 	if _, found := m.subscribers[s.Key()]; !found {
 		return ErrSubscriberDoesNotExist
 	}
 
+	return nil
+}
+
+func (m *manager) checkSubscribersExist(s Subscriber) error {
```
Comment (cosminrentea) [inline server/connector/manager.go]: checkSubscriberExists

File: server/connector/manager.go [hunk 4]
```diff
@@ -106,33 +104,62 @@ func (m *manager) Filter(filters map[string]string) (subscribers []Subscriber) {
...
+	err := m.checkSubscriberDoesNotExists(s)
+	if err != nil {
+		return err
+	}
+
+	m.putSubscriberInMap(s)
+	return m.updateStore(s)
+}
+
+func (m *manager) checkSubscriberDoesNotExists(s Subscriber) error {
```
Comment (cosminrentea) [inline server/connector/manager.go]: checkSubscriberDoesNotExist

File: server/connector/manager.go (hunk 3 above)
Comment (cosminrentea) [inline server/connector/manager.go]: these 2 related functions can be unified into a single func returning a boolean not an error
and their calling functions (upper level) can return the different errors Err..Exists, Err...NotExist

File: server/connector/manager.go [hunk 5]
```diff
@@ -106,33 +104,62 @@ func (m *manager) Filter(filters map[string]string) (subscribers []Subscriber) {
 
 func (m *manager) Add(s Subscriber) error {
 	logger.WithField("subscriber", s).WithField("lock", m.RWMutex).Debug("Add subscriber before locking")
```
Comment (cosminrentea) [inline server/connector/manager.go]: log could be removed

File: server/connector/manager.go [hunk 6]
```diff
@@ -62,13 +62,11 @@ func (m *manager) Create(topic protocol.Path, params router.RouteParams) (Subscr
...
-
 	if m.Exists(key) {
 		logger.WithField("key", key).Debug("Create key exists already")
 		return nil, ErrSubscriberExists
 	}
 
-	logger.Debug("Create  newSubscriber")
 	s := NewSubscriber(topic, params, 0)
 
 	logger.WithField("subscriber", s).Debug("Create  newSubscriber created")
```
Comment (cosminrentea) [inline server/connector/manager.go]: log message can become more clear

File: client/mocks_client_gen_test.go [hunk 7]
```diff
@@ -5,7 +5,7 @@ package client
 
 import (
 	gomock "github.com/golang/mock/gomock"
-	
+
```
Comment (cosminrentea) [inline client/mocks_client_gen_test.go]: minor: revert the mocks (3 files)

**Pull request merged**